src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala

```diff
@@ -0,0 +1,153 @@
+
+    Seq(testGroup, testPolicy).foreach { target =>
+      val mockAccessPolicyDAO = mock[AccessPolicyDAO]
+      val mockDirectoryDAO = mock[DirectoryDAO]
+      val mockGoogleDirectoryDAO = mock[GoogleDirectoryDAO]
+      val ge = new GoogleExtensions(mockDirectoryDAO, mockAccessPolicyDAO, mockGoogleDirectoryDAO, null, null, googleServicesConfig, petServiceAccountConfig)
+
+      target match {
+        case g: BasicWorkbenchGroup =>
+          when(mockDirectoryDAO.loadGroup(g.id)).thenReturn(Future.successful(Option(testGroup)))
+          when(mockDirectoryDAO.updateSynchronizedDate(g.id)).thenReturn(Future.successful(()))
+        case p: AccessPolicy =>
+          when(mockAccessPolicyDAO.loadPolicy(p.id)).thenReturn(Future.successful(Option(testPolicy)))
+          when(mockDirectoryDAO.updateSynchronizedDate(p.id)).thenReturn(Future.successful(()))
+      }
+
+      val subGroups = Seq(inSamSubGroup, inGoogleSubGroup, inBothSubGroup)
+      subGroups.foreach { g => when(mockDirectoryDAO.loadSubjectEmail(g.id)).thenReturn(Future.successful(Option(g.email))) }
+
+      val added = Seq(inSamSubGroup.email.value, ge.toProxyFromUser(inSamUserId.value))
+      val removed = Seq(inGoogleSubGroup.email.value, ge.toProxyFromUser(inGoogleUserId.value))
+
+      when(mockGoogleDirectoryDAO.listGroupMembers(target.email)).thenReturn(Future.successful(Option(Seq(ge.toProxyFromUser(inGoogleUserId.value), ge.toProxyFromUser(inBothUserId.value), inGoogleSubGroup.email.value, inBothSubGroup.email.value, removeError))))
+      added.foreach { email => when(mockGoogleDirectoryDAO.addMemberToGroup(target.email, WorkbenchUserEmail(email))).thenReturn(Future.successful(())) }
+      removed.foreach { email => when(mockGoogleDirectoryDAO.removeMemberFromGroup(target.email, WorkbenchUserEmail(email))).thenReturn(Future.successful(())) }
+
+      val addException = new Exception("addError")
+      when(mockGoogleDirectoryDAO.addMemberToGroup(target.email, WorkbenchUserEmail(ge.toProxyFromUser(addError.value)))).thenReturn(Future.failed(addException))
+
+      val removeException = new Exception("removeError")
+      when(mockGoogleDirectoryDAO.removeMemberFromGroup(target.email, WorkbenchUserEmail(removeError))).thenReturn(Future.failed(removeException))
+
+      val results = runAndWait(ge.synchronizeGroupMembers(target.id))
+
+      results.groupEmail should equal(target.email)
+      results.items should contain theSameElementsAs (
+        added.map(SyncReportItem("added", _, None)) ++
+          removed.map(SyncReportItem("removed", _, None)) ++
+          Seq(
+            SyncReportItem("added", ge.toProxyFromUser(addError.value), Option(ErrorReport(addException))),
+            SyncReportItem("removed", removeError, Option(ErrorReport(removeException)))))
+
+      added.foreach { email => verify(mockGoogleDirectoryDAO).addMemberToGroup(target.email, WorkbenchUserEmail(email)) }
```

Comment:
No need to verify behavior that is already stubbed. If the code cares about the value being returned from the stub (defaults to `null` in this case) then the test will fail if the stubbing is wrong. If you weren't stubbing, you would need the `verify` but you almost always don't need both.

Reply:
I think it is required here. I understand that if it is subbed _and_ called probably no need to verify but another passing case would be that it is stubbed _and not_ called. The test is to make sure that we call add on the right members and not on the wrong members. So in this case I don't need to verify unstubbed functions are not called.

Reply:
Yes, you're right. The test fails without the stubbing but only if it's called. The practice of doing only one of stubbing or verifying assumes that the stubbed return value is vital to some other assertion which isn't the case here.

Since it is pure side effect _and_ the return value doesn't depend on the arguments _and_ you're verifying that the right calls are made, you can slightly simplify the stubbing with blanket declarations:
```
when(mockGoogleDirectoryDAO.addMemberToGroup(any[WorkbenchGroupEmail], any[WorkbenchEmail])).thenReturn(Future.successful(()))
when(mockGoogleDirectoryDAO.removeMemberFromGroup(any[WorkbenchGroupEmail], any[WorkbenchEmail])).thenReturn(Future.successful(()))
```
I like this but not enough to argue passionately about it so it's up to you which style you prefer.

Reply:
thanks, I was trying to figure out how to do that